Improve performance of blocksdb during IBD

ptschip opened the pull request.

ptschip committed:
Add writeoptions.sync and set to false

This was previously not defined but is assumed to be false. Set it to
false to make sure it is set correctly.

ptschip committed:
Write block and undo files asyncronously when the chain is not synced.

This dramatically improves the performance of the blocksdb sync process.

sickpig reviewed (commented):
Two quick questions: 

1) what's the gain obtained with fsync turned off during IBD? 
2) did you try to simulate a unclean termination of bitcoind/qt during IBD to see if in the next restart the IBD process restart nicely?

ptschip commented:
> Two quick questions:
> 
> 1. what's the gain obtained with fsync turned off during IBD?
> 2. did you try to simulate a unclean termination of bitcoind/qt during IBD to see if in the next restart the IBD process restart nicely?

1) Ran tests as follows doing IBD for the first 300,000 blocks:
- with this PR:  28 minutes
- with current dev: 63 minutes

2) did various abrupt terminations and they all were fine. Although it did point out that we should
  probably flush the blockindex at various intervals throughout IBD because we just end up having to 
  restart from scratch if we hadn't had a dbcache flush prior to termination. But that's beyond the scope of this PR...i'll work on that issue separately since it already is a current problem.

sickpig commented:
> > Two quick questions:
> > 
> > 1. what's the gain obtained with fsync turned off during IBD?
> > 2. did you try to simulate a unclean termination of bitcoind/qt during IBD to see if in the next restart the IBD process restart nicely?
> 
> 1. Ran tests as follows doing IBD for the first 300,000 blocks:
> 
> * with this PR:  28 minutes
> * with current dev: 63 minutes

this is a great perf increase here. 

> 1. did various abrupt terminations and they all were fine. Although it did point out that we should
>    probably flush the blockindex at various intervals throughout IBD because we just end up having to
>    restart from scratch if we hadn't had a dbcache flush prior to termination. But that's beyond the scope of this PR...i'll work on that issue separately since it already is a current problem.

agreed on both points.

ptschip committed:
Flush the blockdb databases before flushing the blockindex

We have to make sure all blocks and undo files are written to disk
before commiting the blockindex.

ptschip committed:
Make small fix to compactblocks_2.py
